example/src/main/scala/tamer/example/S3Generalized.scala

```diff
@@ -0,0 +1,90 @@
+package tamer.example
+
+import eu.timepit.refined._
+import eu.timepit.refined.auto._
+import eu.timepit.refined.boolean.{And, Or}
+import eu.timepit.refined.collection.{Forall, NonEmpty}
+import eu.timepit.refined.string.{IPv4, Uri}
+import software.amazon.awssdk.regions.Region
+import tamer.TamerError
+import tamer.config.Config
+import tamer.kafka.Kafka
+import tamer.s3.{Keys, KeysR, Setup}
+import zio.blocking.Blocking
+import zio.clock.Clock
+import zio.duration.durationInt
+import zio.s3.{InvalidCredentials, S3, S3Credentials}
+import zio.stream.{Transducer, ZTransducer}
+import zio.{ExitCode, Has, Layer, Queue, UIO, URIO, ZIO}
+
+import java.net.URI
+import scala.concurrent.duration.{FiniteDuration => FD}
+import scala.util.hashing.MurmurHash3.stringHash
+
+final case class Line(str: String)
+final case class LastProcessedNumber(number: Long)
+
+object S3Generalized extends zio.App {
+  val mkS3Layer: ZIO[Blocking, InvalidCredentials, Layer[RuntimeException, S3 with Kafka]] =
+    S3Credentials.fromAll.map { s3Credentials =>
+      val kafkaState: Config.KafkaState = Config.KafkaState("state-topic", "groupid", "clientid")
+      val kafkaSink: Config.KafkaSink   = Config.KafkaSink("sink-topic")
+      val hostList                      = refineV[NonEmpty And Forall[IPv4 Or Uri]](List("localhost:9092"))
+      val kafkaConfigLayer: Layer[String, Has[Config.Kafka]] =
+        ZIO.fromEither(hostList.map(hl => Config.Kafka(hl, "http://localhost:8081", FD(10, "seconds"), 50, kafkaSink, kafkaState))).toLayer
+      val kafkaLayer: Layer[TamerError, Kafka] = kafkaConfigLayer.mapError(e => TamerError(e)) >>> Kafka.live
+      zio.s3.live(Region.AF_SOUTH_1, s3Credentials, Some(new URI("http://localhost:9000"))) ++ kafkaLayer
+    }
+
+  private val myTransducer: Transducer[Nothing, Byte, Line] =
+    ZTransducer.utf8Decode >>> ZTransducer.splitLines.map(Line)
+
+  private final def getNextNumber(
+      keysR: KeysR,
+      afterwards: LastProcessedNumber,
+      prefix: String
+  ): ZIO[Any, Nothing, Option[Long]] = keysR.get.map { keys =>
+    val sortedFileNumbers = keys
+      .map(key => key.stripPrefix(prefix).toLong)
+      .filter(number => afterwards.number < number)
+
+    if (sortedFileNumbers.isEmpty) None else Some(sortedFileNumbers.min)
```

Comment:
Just FYI you could decide to use scala-collection-compat if you're planning on using collection features of 2.13+ that are not available in earlier

https://github.com/scala/scala-collection-compat/blob/42f7cb266ccc8c61b2f6d68083bad0c141b2c155/compat/src/main/scala-2.11_2.12/scala/collection/compat/PackageShared.scala#L260

Which reminds me we need to scrutinise what we have atm as it relies on functions available in 10+ and 11+. Noticed when I compiled with my default adopt jdk 1.8

Reply:
you are talking about jdk > 8 used by 2.13? For example?

Reply:
nope, pure jdk added methods:

- [`localizedBy`](https://docs.oracle.com/javase/10/docs/api/java/time/format/DateTimeFormatter.html#localizedBy(java.util.Locale)) since JDK10 used in https://github.com/laserdisc-io/tamer/blob/7ac5a367820cd31df0d9a083b2103944c617411d/s3/src/test/scala/tamer/s3/DateParsingSpec.scala#L14 and https://github.com/laserdisc-io/tamer/blob/7ac5a367820cd31df0d9a083b2103944c617411d/s3/src/test/scala/tamer/s3/DateParsingSpec.scala#L22
- [`stripLeading`](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/lang/String.html#stripLeading()) since JDK11 used in https://github.com/laserdisc-io/tamer/blob/c7d379c11e325f312e08c5c6b927cca5e6839df2/doobie/src/main/scala/tamer/db/Setup.scala#L43